### PR TITLE
Add CreateBlockBlobWithData and CreateBlockBlobFromReader

### PR DESCRIPTION
**Pull request by tomclegg**

### CODE REVIEW

**Comment by ahmetalpbalkan** [comment]
@tomclegg thanks for the pull request. We used to have a version of this method that was accepting data as you submitted and we ended up deleting that.

This sort of gives the impression that any size of data in `[]byte` or `io.Reader` would work with these calls; whereas, that's not true. This API call has a 16 or 64 MB limit (or something like that) and bigger files should be chunked into separate requests that look like: 

    CreateBlob --> PutBlock(data) --> PutBlock(data) -->  ... --> PutBlockList(blocks)

however the initial call to CreateBlob also accepts data, but this creates a situation where user is not in control of the block ID this data is uploaded to, therefore manipulating the blob later on using PutBlock/PutBlockList calls become harder. That's why I try to avoid sending data in this CreateBlob call...

We have a plan to implement #208  and that would happily accommodate `[]byte`/`io.Reader`. Does it make sense to you as well to implement this contribution for that `PutBlockBlob` in #208 that would work with any size of contents, instead of this initial call?

**Comment by tomclegg** [comment]
My use case involves creating a 64 MiB block atomically, and I don't care what the block IDs are because my blocks are immutable, so PutBlob seems like the right API. PutBlock, on the other hand, means uploading 16x 4MiB blocks: forking the SDK to use the more appropriate API seemed easier and neater.

If I'm following #129 and #208 correctly, the plan is to add "upload small blobs in one go, split bigger blobs into blocks and upload with optional concurrency." That sounds useful too.

It seems to me that even if we have `blobutil.PutBlockBlob` it will still make sense to have a `PutBlob` that just calls the corresponding REST API, which is what I was aiming for here. So I think this solves a different problem than #208 (but might make #208 easier to write).

Re size restriction: PutBlock seems to ignore the size and let the server decide whether it's 413, so I figured PutBlob should use the same approach. Ideally the server would tell us the maximum size so we don't end up imposing different limits in the SDK and the actual API. (Does it?) I'd be happy to add a size check if you like.

**Comment by ahmetalpbalkan** [comment]
@tomclegg giving a second thought your argument makes sense. This is a valid API call that users should be allowed to use. `blobutil.PutBlockBlob` probably will end up using this call with no data to have full control of block IDs (they all have to be the same length).

We shouldn't check any sizes on the client side as you said.

**Comment by ahmetalpbalkan** [comment]
@tomclegg however, I think it'd be useful to document the max sizes on these methods as godoc comments. Do you mind doing that? Also IMHO `..WithData` might be a bit obsolete because `[]byte` can be easily made a reader by `bytes.NewReader(data)`.

**Comment by tomclegg** [comment]
Should I rename CreateBlockBlobWithData to PutBlockBlob or PutBlob? (I named it to match the existing SDK function, but following the REST API name would also seems reasonable. FWIW, when I first tried to find it in the SDK, I looked for PutBlob.)

**Comment by tomclegg** [comment]
@ahmetalpbalkan yes, ...WithData is kind of sugary, I'll remove it.

**Comment by ahmetalpbalkan** [comment]
CreateBlockBlobWithReader is better IMO. PutBlockBlob/PutBlob is the REST API docs terminology but I don't necessarily agree it makes it clear that this is a size-limited endpoint that should be used to create first chunk of the blob.

File: storage/blob_test.go [hunk 1]
```diff
@@ -389,6 +390,41 @@ func (s *StorageBlobSuite) TestGetBlobRange(c *chk.C) {
...
+
+func (s *StorageBlobSuite) TestCreateBlockBlobFromReaderWithShortData(c *chk.C) {
+	cli := getBlobClient(c)
+	cnt := randContainer()
+	c.Assert(cli.CreateContainer(cnt, ContainerAccessTypePrivate), chk.IsNil)
+	defer cli.deleteContainer(cnt)
+
+	name := randString(20)
+	data := randBytes(8888)
+	err := cli.CreateBlockBlobFromReader(cnt, name, 9999, bytes.NewReader(data))
```
Comment (ahmetalpbalkan) [inline storage/blob_test.go]: this test is now pretty much the same thing as the other test, right? Also you're saying I'm sending 9999 bytes but `data` is 8888 bytes.

Comment (tomclegg) [inline storage/blob_test.go]: Yes, up to this line the test is just like the previous one. But the previous one only tests a "success" case. This one tests whether an upload error ("we said we were uploading a 9999-byte blob, but we provided fewer than 9999 bytes of data") causes `CreateBlock` to return an error (so our application knows something went wrong) and avoids storing the truncated data in the container (so we don't have to wonder every time our application calls `GetBlob` whether we're getting back bogus data due to a previous upload mishap).

Perhaps `...WithTruncatedData` would be a better hint than `...WithShortData` about what's being tested?

Comment (ahmetalpbalkan) [inline storage/blob_test.go]: ah sorry didn't see the `Not(..)` part.

**Comment by ahmetalpbalkan** [comment]
LGTM. Thanks @tomclegg.

**Pull request merged**